### PR TITLE
Allow range finder to be used

### PR DESCRIPTION
**Pull request by andyp1per**
Currently ```EK2_RNG_USE_HGT``` will not use the rangefinder for height calculations - we always use the barometer. This change re-enables the use of the rangefinder close to the ground.

This fix survives @rmackay9 's test in https://github.com/ArduPilot/ardupilot/pull/12250

### CODE REVIEW

**Comment by andyp1per** [comment]
@anbello this worked for me in SITL, no idea whether it is actually the right thing

**Comment by andyp1per** [comment]
The fact that I have enabled some code and commented out other pieces and the tests all pass is a little disconcerting!

**Comment by anbello** [comment]
This commit should fix the fact that even with EK2_RNG_USE_HGT set the rangefinder is not used close to ground with EK2_ALT_SOURCE != 1. I will test it asap.

I would like to know if you tested also the other issue that there is no switch to baro if the copter goes over the limit RNGFND_MAX_CM with rangefinder set as primary alt source.

**Comment by andyp1per** [comment]
It switches to baro as long as you do *not* set EK2_ALT_SOURCE to rangefinder, which I think is what you want

**Comment by anbello** [comment]
Don't you think that even with EK2_ALT_SOURCE set to rangefinder there should be a switch to baro when we go outside the limits? For safety reason.

**Comment by andyp1per** [comment]
@anbello yeah that might work the same way looking at the code. Does it not work today?

**Comment by anbello** [comment]
>  Does it not work today?

Do you mean if I tested today? Sorry but I haven't the time.

**Comment by andyp1per** [comment]
No I mean I thought you were flying with EK2_ALT_SOURCE anyway, are you not?

**Comment by anbello** [comment]
Yes, near always when I fly indoor, but indoor I don't go ever the limit of rangefinder.
One time I did it outdoor, with EK2_ALT_SOURCE = 1 and  EK2_RNG_USE_HGT = 70 (only way to have rangefinder as primary with actual code) and when I gone over RNGFND_MAX_CM the copter continued to climb even with throttle at center.

**Comment by andyp1per** [comment]
That's my worry!

**Comment by anbello** [comment]
Here a SITL sim of this situation
https://discuss.ardupilot.org/t/lidar-in-althold/51090/13?u=anbello

**Comment by andyp1per** [comment]
So your simulation works for:
```EK2_ALT_SOURCE = 0 EK2_RNG_USE_HGT = 70```
it does not work for 
```EK2_ALT_SOURCE = 1 EK2_RNG_USE_HGT = 70```
☹️

File: libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp [hunk 1]
```diff
@@ -129,7 +129,9 @@ void NavEKF2_core::EstimateTerrainOffset()
 
             // calculate Kalman gain
             float SK_RNG = sq(q0) - sq(q1) - sq(q2) + sq(q3);
+#if 0
```
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp:132]: Range finder should only be used to either directly observe height for the main filter or to estimate the height of terrain wrt the EKF origin, not both. Any change to range finder use logic needs to be coordinated between NavEKF2_core::selectHeightForFusion() and NavEKF2_core::EstimateTerrainOffset() to ensure that this requirement is not violated.

File: libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp [hunk 2]
```diff
@@ -942,7 +942,7 @@ void NavEKF2_core::selectHeightForFusion()
     if (extNavUsedForPos) {
         // always use external vision as the height source if using for position.
         activeHgtSource = HGT_SOURCE_EXTNAV;
-    } else if (_rng && ((frontend->_useRngSwHgt > 0) && (frontend->_altSource == 1)) && (imuSampleTime_ms - rngValidMeaTime_ms < 500)) {
+    } else if (_rng && (frontend->_useRngSwHgt > 0) && (imuSampleTime_ms - rngValidMeaTime_ms < 500)) {
```
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp:945]: This should be split into two cases. One when frontend->_altSource == 1 which always ruses the range finder as the active height source and one when frontend->_useRngSwHgt > 0  and frontend->_altSource != which switches between the primary height source and range finder using the height and speed criteria.

**Comment by peterbarker** [comment]
> The fact that I have enabled some code and commented out other pieces and the tests all pass is a little disconcerting!

Then fix and augment the tests.  They don't write themselves.

File: libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp (hunk 2 above)
Comment (anbello) [reply libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp:945]: > ... and one when frontend->_useRngSwHgt > 0 and frontend->_altSource != which ...

I Think this should be:
... and one when frontend->_useRngSwHgt > 0 and frontend->_altSource != 1 which ...

File: libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp (hunk 1 above)
Comment (priseborough) [reply libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp:132]: You can see here: https://github.com/ArduPilot/ardupilot/blob/8f7bf2137234ba55c88b4fc978202221fa7ed2a1/libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp#L95 that current behaviour is to inhibit terrain height estimation if range finder is being used as the primary height source and optical flow is being used for navigationand not terrain height estimation.

Comment (priseborough) [reply libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp:132]: I don't understand why you have added a compiler directive to bypass calculation of this term.

Comment (andyp1per) [reply libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp:132]: It was a test to demonstrate the problem. Not suggesting this is the real fix!

Comment (andyp1per) [reply libraries/AP_NavEKF2/AP_NavEKF2_OptFlowFusion.cpp:132]: As discussed, the inhibition logic is failing at the crucial point that the rangefinder data becomes active leading to a very large delta in the terrainState which causes the copter to think it is 15m in the air when it has hit the ground.

**Comment by andyp1per** [comment]
@anbello by the way your use case (EK2_ALT_SOURCE = 1) is by design - we *must* not rely on the baro in these circumstances because inside it could be way off. I suppose the only reasonable thing to do is fence at the max rangefinder range, but not sure how easy that would be.

**Comment by andyp1per** [comment]
I've written a test for this, but its not failing. Whereas a manual test fails. Here is a screen grab of the relevant (bad) transition for a manual test:

image

**Comment by andyp1per** [comment]
@priseborough I have made the changes you suggested and actually now this PR completely reverses #12250 which is clearly completely the opposite of what was intended.
I have added a test that does the equivalent of @rmackay9 s manual test, but it passes - I cannot figure out why. If I run these steps manually using sim_vehicle then it fails.

**Comment by priseborough** [comment]
Please continue discussion in https://github.com/ArduPilot/ardupilot/pull/14375

**Pull request closed without merging**